### PR TITLE
Add array_agg function

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/ArrayAggregationStateFactory.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        public void ensureCapacity(long size)
+        {
+            blockBuilders.ensureCapacity(size);
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            //TODO: Is this doing what it should do?
+            return blockBuilders.sizeOf();
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/ArrayAggregationStateFactory.java]: @cberner Let's talk about this one tomorrow.

**Comment by haozhun** [comment]
@cberner  Also, let's talk about whether `FixedWidthBlockBuilder` should be created at all.

**Comment by haozhun** [comment]
Ready for review

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    {
+        Type type = types.get("T");
+        Signature signature = new Signature(NAME,
+                parameterizedTypeName("array", type.getTypeSignature()),
+                type.getTypeSignature());
+        InternalAggregationFunction aggregation = generateAggregation(type);
+        return new FunctionInfo(signature, getDescription(), aggregation);
+    }
+
+    protected InternalAggregationFunction generateAggregation(Type type)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java]: you can make this `private`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                outputType,
+                false);
+
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, outputType, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type value)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(NULLABLE_INPUT_CHANNEL, value), new ParameterMetadata(BLOCK_INDEX));
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java]: The spec says that `array_agg` skips NULLs and returns NULL if it read zero rows

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        BlockBuilder stateBlockBuilder = state.getBlockBuilder();
+        BlockBuilder otherStateBlockBuilder = otherState.getBlockBuilder();
+        if (otherStateBlockBuilder == null) {
+            return;
+        }
+        if (stateBlockBuilder == null) {
+            state.setBlockBuilder(otherStateBlockBuilder);
+            return;
+        }
+        int otherStateBlockBuilderPositionCount = otherStateBlockBuilder.getPositionCount();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java]: this name is kind of long. I would name it `positionCount`

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestArrayAggregation.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static com.facebook.presto.type.TypeUtils.buildStructuralSlice;
+import static org.testng.Assert.assertNotNull;
+
+public class TestArrayAggregation
+{
+    private static final MetadataManager metadata = new MetadataManager();
+
+    @Test
+    public void testAllRegistered()
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestArrayAggregation.java]: you can remove this test. It was for a very specific bug that was fixed a while ago

**Comment by cberner** [comment]
seemed like the right approach. Assign back to me when you've change it to meet the spec and I'll finish reviewing it.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java (hunk 3 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java]: Just noticed that this is also the default for all aggregate functions in presto. https://prestodb.io/docs/current/functions/aggregate.html

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/ArrayAggregationState.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package com.facebook.presto.operator.aggregation.state;
+
+import com.facebook.presto.spi.block.BlockBuilder;
+
+@AccumulatorStateMetadata(stateFactoryClass = ArrayAggregationStateFactory.class, stateSerializerClass = ArrayAggregationStateSerializer.class)
+public interface ArrayAggregationState
+    extends AccumulatorState
+{
+    BlockBuilder getBlockBuilder();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/ArrayAggregationState.java]: add a blank line after getter

**Comment by cberner** [comment]
looks good

**Pull request merged**